core/inlooptool.py

```diff
@@ -966,12 +1265,165 @@ def add_build_year_to_surface(self, file_path, field_name="bouwjaar"):
         surfaces = None
         print("... done")
         return
+    
+    def _save_to_gpkg_test(self, file_path): #TO DO: weghalen, is alleen voor testen
+        print("Preparing template gpkg")
+        output_gpkg = r"C:\Users\ruben.vanderzaag\Documents\Github\bgt-inlooptool\QGIS_plugin\bgtinlooptool\style\output_bgtinlooptool_testing_rekeninstellingen.gpkg"
 
-    def _write_to_disk(self, file_path):
-        """Copy self.mem_database to file_path"""
-        self.out_db = GPKG_DRIVER.CopyDataSource(self.mem_database, file_path)
+        print("Testen van output wegschrijven!!")
+        self.out_db = GPKG_DRIVER.CopyDataSource(self.mem_database, output_gpkg)
         self.out_db = None
+    
+    def _save_to_gpkg(self, file_path):
```

Comment:
Je vraag was of hier een beter performance mogelijk is. 

Wat me meteen opvalt is dat er 5x dezelfde methode wordt aangeroepen, waarin vervolgens de geopackage wordt geopend. Het lijkt me handiger om in deze methode de geopackage eenmalig (in een context manager, zie andere comment) te openen, om vervolgens de aanpassingen te doen. Of dit veel verschil maakt weet ik niet, maar netter is het sowieso. gpt gaf als voorbeeld:

def _save_to_gpkg(self, file_path):
    print("Preparing template gpkg")
    template_gpkg = r"C:\Users\ruben.vanderzaag\Documents\Github\bgt-inlooptool\QGIS_plugin\bgtinlooptool\style\template_output15.gpkg"
    output_gpkg = r"C:\Users\ruben.vanderzaag\Documents\Github\bgt-inlooptool\QGIS_plugin\bgtinlooptool\style\output_bgtinlooptool.gpkg"
    self.copy_and_rename_file(template_gpkg, output_gpkg)

    print("Saving layers to gpkg")
    layers = [
        (PIPES_TABLE_NAME, "3. GWSW leidingen"),
        ("bgt_inlooptabel", "4. BGT inlooptabel"),
        (SURFACES_TABLE_NAME, "5. BGT oppervlakken"),
        (STATISTICS_TABLE_NAME, "6. Statistieken"),
        (SETTINGS_TABLE_NAME, "7. Rekeninstellingen")
    ]

    with self.open_gpkg(file_path) as dst_gpkg:
        for db_layer, gpkg_layer in layers:
            print(f"Saving {gpkg_layer} layer in gpkg")
            self._write_to_disk(dst_gpkg, db_layer, gpkg_layer)
            self.track_changes(file_path)
            
    print("All layers saved successfully.")

def copy_and_rename_file(self, original_file_path, new_file_path):
    try:
        with open(original_file_path, 'rb') as original_file:
            content = original_file.read()
        with open(new_file_path, 'wb') as new_file:
            new_file.write(content)
        print(f"File copied from {original_file_path} to {new_file_path}")
    except FileNotFoundError:
        print(f"The file {original_file_path} does not exist.")
    except PermissionError:
        print(f"Permission denied. Unable to copy {original_file_path} to {new_file_path}.")
    except Exception as e:
        print(f"An error occurred: {e}")

def _write_to_disk(self, dst_gpkg, db_layer_name, dst_layer_name):
    db_layer = self.mem_database.GetLayerByName(db_layer_name)
    if db_layer is None:
        raise ValueError(f"Layer '{db_layer_name}' not found in memory database.")
    
    dst_layer = dst_gpkg.GetLayerByName(dst_layer_name)
    if dst_layer is None:
        raise ValueError(f"Layer '{dst_layer_name}' not found in destination GeoPackage.")
    
    layer_defn = db_layer.GetLayerDefn()
    dst_layer_defn = dst_layer.GetLayerDefn()
    
    if layer_defn.GetFieldCount() != dst_layer_defn.GetFieldCount():
        print(f"Warning: Source and destination layers have different field counts: {layer_defn.GetFieldCount()} vs {dst_layer_defn.GetFieldCount()}")

    field_mapping = {dst_layer_defn.GetFieldDefn(i).GetName(): j
                     for i in range(dst_layer_defn.GetFieldCount())
                     for j in range(layer_defn.GetFieldCount())
                     if dst_layer_defn.GetFieldDefn(i).GetName() == layer_defn.GetFieldDefn(j).GetName()}

    for feature in db_layer:
        dst_feature = ogr.Feature(dst_layer_defn)
        for dst_field_name, src_field_index in field_mapping.items():
            dst_feature.SetField(dst_field_name, feature.GetField(src_field_index))
        geom = feature.GetGeometryRef()
        if geom:
            dst_feature.SetGeometry(geom.Clone())
        dst_layer.CreateFeature(dst_feature)
        dst_feature = None

    print(f"Layer '{dst_layer_name}' saved successfully.")

@contextlib.contextmanager
def open_gpkg(self, file_path):
    dst_gpkg = GPKG_DRIVER.Open(file_path, 1)
    if dst_gpkg is None:
        raise ValueError(f"Could not open GeoPackage '{file_path}' for writing.")
    try:
        yield dst_gpkg
    finally:
        dst_gpkg = None

Reply:
```python

def _save_to_gpkg(self, file_path):
    print("Preparing template gpkg")
    template_gpkg = r"C:\Users\ruben.vanderzaag\Documents\Github\bgt-inlooptool\QGIS_plugin\bgtinlooptool\style\template_output15.gpkg"
    output_gpkg = r"C:\Users\ruben.vanderzaag\Documents\Github\bgt-inlooptool\QGIS_plugin\bgtinlooptool\style\output_bgtinlooptool.gpkg"
    self.copy_and_rename_file(template_gpkg, output_gpkg)

    print("Saving layers to gpkg")
    layers = [
        (PIPES_TABLE_NAME, "3. GWSW leidingen"),
        ("bgt_inlooptabel", "4. BGT inlooptabel"),
        (SURFACES_TABLE_NAME, "5. BGT oppervlakken"),
        (STATISTICS_TABLE_NAME, "6. Statistieken"),
        (SETTINGS_TABLE_NAME, "7. Rekeninstellingen")
    ]

    with self.open_gpkg(file_path) as dst_gpkg:
        for db_layer, gpkg_layer in layers:
            print(f"Saving {gpkg_layer} layer in gpkg")
            self._write_to_disk(dst_gpkg, db_layer, gpkg_layer)
            self.track_changes(file_path)
            
    print("All layers saved successfully.")

def copy_and_rename_file(self, original_file_path, new_file_path):
    try:
        with open(original_file_path, 'rb') as original_file:
            content = original_file.read()
        with open(new_file_path, 'wb') as new_file:
            new_file.write(content)
        print(f"File copied from {original_file_path} to {new_file_path}")
    except FileNotFoundError:
        print(f"The file {original_file_path} does not exist.")
    except PermissionError:
        print(f"Permission denied. Unable to copy {original_file_path} to {new_file_path}.")
    except Exception as e:
        print(f"An error occurred: {e}")

def _write_to_disk(self, dst_gpkg, db_layer_name, dst_layer_name):
    db_layer = self.mem_database.GetLayerByName(db_layer_name)
    if db_layer is None:
        raise ValueError(f"Layer '{db_layer_name}' not found in memory database.")
    
    dst_layer = dst_gpkg.GetLayerByName(dst_layer_name)
    if dst_layer is None:
        raise ValueError(f"Layer '{dst_layer_name}' not found in destination GeoPackage.")
    
    layer_defn = db_layer.GetLayerDefn()
    dst_layer_defn = dst_layer.GetLayerDefn()
    
    if layer_defn.GetFieldCount() != dst_layer_defn.GetFieldCount():
        print(f"Warning: Source and destination layers have different field counts: {layer_defn.GetFieldCount()} vs {dst_layer_defn.GetFieldCount()}")

    field_mapping = {dst_layer_defn.GetFieldDefn(i).GetName(): j
                     for i in range(dst_layer_defn.GetFieldCount())
                     for j in range(layer_defn.GetFieldCount())
                     if dst_layer_defn.GetFieldDefn(i).GetName() == layer_defn.GetFieldDefn(j).GetName()}

    for feature in db_layer:
        dst_feature = ogr.Feature(dst_layer_defn)
        for dst_field_name, src_field_index in field_mapping.items():
            dst_feature.SetField(dst_field_name, feature.GetField(src_field_index))
        geom = feature.GetGeometryRef()
        if geom:
            dst_feature.SetGeometry(geom.Clone())
        dst_layer.CreateFeature(dst_feature)
        dst_feature = None

    print(f"Layer '{dst_layer_name}' saved successfully.")

@contextlib.contextmanager
def open_gpkg(self, file_path):
    dst_gpkg = GPKG_DRIVER.Open(file_path, 1)
    if dst_gpkg is None:
        raise ValueError(f"Could not open GeoPackage '{file_path}' for writing.")
    try:
        yield dst_gpkg
    finally:
        dst_gpkg = None
        
```

Reply:
Toegevoegd aan script. Werkt nu goed. Lijkt iets sneller, maar is een kwestie van seconden.